test-suite/shared/src/test/scala/org/scalajs/testsuite/niocharset/BaseCharsetTest.scala

```diff
@@ -30,6 +30,57 @@ class BaseCharsetTest(val charset: Charset) {
   protected def testDecode(in: ByteBuffer)(
       outParts: OutPart[CharBuffer]*): Unit = {
 
+    def testDecodeVsSteppedDecode(malformedAction: CodingErrorAction,
+        unmappableAction: CodingErrorAction) = {
+
+      val in1 = in.duplicate
+      val in2 = in.duplicate
+
+      in2.limit(in2.position)
+
+      val d1 = charset.newDecoder()
+      d1.onMalformedInput(malformedAction)
+      d1.onUnmappableCharacter(unmappableAction)
```

Comment:
You'll probably want to abstract away this 3-line incantation in a separate `def newDecoder(malformedAction, unmappableAction)` inside `testDecode` (or even as `private def` in `BaseCharsetTest`). This incantation is now repeated 3 times (twice in your new code, and once in `testOneConfig`).